swap/swap.go

```diff
@@ -1,98 +1,451 @@
+	s.balances[peer] += amount
 	//save the new balance to the state store
-	peerBalance := s.balances[peer.ID()]
-	err = s.stateStore.Put(peer.ID().String(), &peerBalance)
+	peerBalance := s.balances[peer]
+	err := s.store.Put(balanceKey(peer), &peerBalance)
+	if err != nil {
+		return 0, fmt.Errorf("error while storing balance for peer %s", peer.String())
+	}
+	log.Debug("balance for peer after accounting", "peer", peer.String(), "balance", strconv.FormatInt(peerBalance, 10))
+	return peerBalance, err
+}
 
-	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
-	return err
+// loadBalance loads balances from the state store (persisted)
+func (s *Swap) loadBalance(peer enode.ID) (err error) {
+	var peerBalance int64
+	//only load if the current instance doesn't already have this peer's
+	//balance in memory
+	if _, ok := s.balances[peer]; !ok {
+		err = s.store.Get(balanceKey(peer), &peerBalance)
+		s.balances[peer] = peerBalance
+	}
+	return
 }
 
-//GetPeerBalance returns the balance for a given peer
-func (swap *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
-	swap.lock.RLock()
-	defer swap.lock.RUnlock()
-	if p, ok := swap.balances[peer]; ok {
-		return p, nil
+// sendCheque sends a cheque to peer
+func (s *Swap) sendCheque(peer enode.ID) error {
+	swapPeer, err := s.getPeer(peer)
+	if err != nil {
+		return fmt.Errorf("error while getting peer: %s", err.Error())
+	}
+	cheque, err := s.createCheque(peer)
+	if err != nil {
+		return fmt.Errorf("error while creating cheque: %s", err.Error())
+	}
+
+	log.Info("sending cheque", "serial", cheque.ChequeParams.Serial, "amount", cheque.ChequeParams.Amount, "beneficiary", cheque.Beneficiary, "contract", cheque.Contract)
+	s.cheques[peer] = cheque
+
+	err = s.store.Put(sentChequeKey(peer), &cheque)
+	if err != nil {
+		return fmt.Errorf("error while storing the last cheque: %s", err.Error())
+	}
+
+	emit := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+
+	// reset balance;
+	err = s.resetBalance(peer, int64(cheque.Amount))
+	if err != nil {
+		return err
 	}
-	return 0, errors.New("Peer not found")
+
+	return swapPeer.Send(context.Background(), emit)
 }
 
-//load balances from the state store (persisted)
-func (s *Swap) loadState(peer *protocols.Peer) (err error) {
-	var peerBalance int64
-	peerID := peer.ID()
+// Create a Cheque structure emitted to a specific peer as a beneficiary
+// The serial and amount of the cheque will depend on the last cheque and current balance for this peer
+// The cheque will be signed and point to the issuer's contract
+func (s *Swap) createCheque(peer enode.ID) (*Cheque, error) {
+	var cheque *Cheque
+	var err error
+
+	swapPeer, err := s.getPeer(peer)
+	if err != nil {
+		return nil, fmt.Errorf("error while getting peer: %s", err.Error())
+	}
+	beneficiary := swapPeer.beneficiary
+
+	peerBalance := s.balances[peer]
+	// the balance should be negative here, we take the absolute value:
+	honey := uint64(-peerBalance)
```

Comment:
the comment says you want the absolute value, then please use `math.Abs` and don't try to underrun a `uint64`

Reply:
`math.Abs` only accepts `float64` as a parameter.

i'm not convinced with the idea casting twice, and i think the alternative is what we currently have.

Reply:
i still don't understand this. `uint64` of negative value causes an [underflow](https://play.golang.org/p/FFDajfFmZjM).
i don't understand if `peerBalance` should be normally negative or positive, but I would at least add a check here that the value is indeed negative and panic if not

Reply:
I agree, there should be a check if the peerBalance is indeed negative. Just an assumption here can cause an integer overflow resulting a very large number for the honey. I do not think that math.Abs should be used, just an if statement.